services/py-api/src/database/repository/participants_repository.py

```diff
@@ -0,0 +1,54 @@
+from typing import Optional, Type
+
+from bson import ObjectId
+from motor.motor_asyncio import AsyncIOMotorClientSession
+from pydantic import BaseModel
+from pymongo.errors import DuplicateKeyError
+from result import Result, Ok, Err
+from structlog.stdlib import get_logger
+
+from src.database.db_manager import DatabaseManager
+from src.database.model.participant_model import Participant
+from src.database.repository.base_repository import CRUDRepository
+from src.server.exception import DuplicateEmail
+from src.server.schemas.request_schemas.schemas import ParticipantRequestBody
+
+LOG = get_logger()
+
+
+class ParticipantsRepository(CRUDRepository):
+    # TODO: Implement the rest of the methods
+    def __init__(self, db_manager: DatabaseManager, collection: str) -> None:
+        self._client = db_manager.client
+        self._collection = self._client[db_manager.DB_NAME][collection]
+
+    async def fetch_by_id(self, obj_id: str) -> Result:
+        raise NotImplementedError()
+
+    async def fetch_all(self) -> Result:
+        raise NotImplementedError()
+
+    async def update(
+        self, obj_id: str, input_data: BaseModel, session: Optional[AsyncIOMotorClientSession] = None
+    ) -> Result:
+        raise NotImplementedError()
+
+    async def delete(self, obj_id: str, session: Optional[AsyncIOMotorClientSession] = None) -> Result:
+        raise NotImplementedError()
+
+    async def create(
+        self, input_data: ParticipantRequestBody, session: Optional[AsyncIOMotorClientSession] = None
+    ) -> Ok[Participant] | Err[Type[DuplicateEmail]] | Err[Exception]:
```

Comment:
That's a weird return type. You're forcing yourself into mandatory type check if value is of type `Err`.

Reply:
This was because of mypy i think.

Reply:
Fixed in 1a00dd8c6f9016c7770fd6e17ffc6b8195dc8ce2